packages/backend/src/index.ts

```diff
@@ -58,38 +60,31 @@ import badges from './plugins/badges';
 import jenkins from './plugins/jenkins';
 import permission from './plugins/permission';
 import { PluginEnvironment } from './types';
-import { ServerPermissionClient } from '@backstage/plugin-permission-node';
+import { rootDependencies } from './rootContext';
+import { permissionAuthorizerDep } from '@backstage/plugin-permission-common';
 
-function makeCreateEnv(config: Config) {
+function makeCreateEnv(config: Config, applicationContext: ApplicationContext) {
   const root = getRootLogger();
-  const reader = UrlReaders.default({ logger: root, config });
-  const discovery = SingleHostDiscovery.fromConfig(config);
-  const tokenManager = ServerTokenManager.fromConfig(config, { logger: root });
-  const permissions = ServerPermissionClient.fromConfig(config, {
-    discovery,
-    tokenManager,
-  });
-  const databaseManager = DatabaseManager.fromConfig(config);
-  const cacheManager = CacheManager.fromConfig(config);
-  const taskScheduler = TaskScheduler.fromConfig(config);
-
-  root.info(`Created UrlReader ${reader}`);
 
   return (plugin: string): PluginEnvironment => {
     const logger = root.child({ type: 'plugin', plugin });
+    const databaseManager = applicationContext.get(databaseManagerDep);
     const database = databaseManager.forPlugin(plugin);
-    const cache = cacheManager.forPlugin(plugin);
-    const scheduler = taskScheduler.forPlugin(plugin);
+    const cache = applicationContext.get(cacheManagerDep).forPlugin(plugin);
```

Comment:
So resuming the `forPlugin` discussion here, what is really the benefit of this as opposed to what we had previously? In both situations the cache manager implementation is controlled by the app

Reply:
For this it doesn't do anything, it is still the same implementation since the factory is using the same code as the deleted line was. It does give the possibility to swap the implementation of that cachemanager across the application by modifying the `rootContext` construction code, and that happens across the application only if the instance is taken from the container.